libraries/ESP8266WebServer/examples/FSBrowser/FSBrowser.ino

```diff
@@ -226,29 +302,324 @@ void handleFileList() {
+   If it's a file, delete it.
+   If it's a folder, delete all nested contents first then the folder itself
+*/
+void deleteRecursive(String path) {
+  File file = fileSystem->open(path, "r");
+  bool isDir = file.isDirectory();
+  file.close();
+
+  // If it's a plain file, delete it
+  if (!isDir) {
+    fileSystem->remove(path);
+    return;
+  }
+
+  // Otherwise delete its contents first
+  Dir dir = fileSystem->openDir(path);
+
+  while (dir.next()) {
+    deleteRecursive(path + "/" + dir.fileName());
+  }
+
+  // Then delete the folder itself
+  fileSystem->rmdir(path);
+}
+
+
+/*
+   Handle a file deletion request
+   Operation      | req.responseText
+   ---------------+--------------------------------------------------------------
+   Delete file    | parent of deleted file, or remaining ancestor
+   Delete folder  | parent of deleted folder, or remaining ancestor
+*/
+void handleFileDelete() {
+  if (!fsOK) {
+    return returnFail("FS INIT ERROR");
+  }
+
+  String path = server.arg(0);
+  if (path == "") {
+    return returnFail("BAD ARGS");
+  }
+
+  DBG_OUTPUT_PORT.println(String("handleFileDelete: ") + path);
+  if (path == "/" || !fileSystem->exists(path)) {
+    return returnFail("BAD PATH");
+  }
+  deleteRecursive(path);
+
+  returnOKWithMsg(lastExistingParent(path));
+}
+
+/*
+   Handle a file upload request
+*/
+void handleFileUpload() {
+  if (!fsOK) {
+    return returnFail("FS INIT ERROR");
+  }
+  if (server.uri() != "/edit") {
+    return;
+  }
+  HTTPUpload& upload = server.upload();
+  if (upload.status == UPLOAD_FILE_START) {
+    String filename = upload.filename;
+    // Make sure paths always start with "/"
+    if (!filename.startsWith("/")) {
+      filename = "/" + filename;
+    }
+    DBG_OUTPUT_PORT.println(String("handleFileUpload Name: ") + filename);
+    uploadFile = fileSystem->open(filename, "w");
+    if (!uploadFile) {
+      return returnFail("CREATE FAILED");
+    }
+    DBG_OUTPUT_PORT.println(String("Upload: START, filename: ") + filename);
+  } else if (upload.status == UPLOAD_FILE_WRITE) {
+    if (uploadFile) {
+      size_t bytesWritten = uploadFile.write(upload.buf, upload.currentSize);
+      if (bytesWritten != upload.currentSize) {
+        return returnFail("WRITE FAILED");
+      }
+    }
+    DBG_OUTPUT_PORT.println(String("Upload: WRITE, Bytes: ") + upload.currentSize);
+  } else if (upload.status == UPLOAD_FILE_END) {
+    if (uploadFile) {
+      uploadFile.close();
+    }
+    DBG_OUTPUT_PORT.println(String("Upload: END, Size: ") + upload.totalSize);
+  }
+}
+
+
+/*
+   The "Not Found" handler catches all URI not explicitely declared in code
+   First try to find and return the requested file from the filesystem,
+   and if it fails, return a 404 page with debug information
+*/
+void handleNotFound() {
+  if (!fsOK) {
+    return returnFail("FS INIT ERROR");
+  }
+  String uri = urlDecode(server.uri());
+  if (!handleFileRead(uri)) {
```

Comment:
Prefer direct logic over inverted:
```cpp
if(handleFileRead(uri)
  return;

String message ...
...
return replyNotFound(message);
}
````

Again, please avoid/fix such sequences of concats. Call String::reserve() first with an appropriate value.
And please move the strings to flash.